cupy/_core/dlpack.pyx

```diff
@@ -18,140 +13,244 @@ import warnings
+        device.device_type = kDLCPU
+        device.device_id = 0
+
+    cdef void *dlm_tensor_ptr = stdlib.malloc(
+        sizeof(DLManagedTensorVersioned) if use_versioned
+        else sizeof(DLManagedTensor)
+    )
+    if dlm_tensor_ptr == NULL:
+        raise MemoryError()
+
+    # Note: could coalesce this with the previous allocation in principle
+    cdef int64_t* shape_strides = <int64_t*>stdlib.malloc(
+        ndim * sizeof(int64_t) * 2)
+    if shape_strides == NULL:
+        stdlib.free(dlm_tensor_ptr)
+        raise MemoryError()
+
+    # We need a different setup for versioned/unversioned when it comes to
+    # the context/deleter and additional info in the newer versioned one.
+    if use_versioned:
+        dlm_tensor_ver = <DLManagedTensorVersioned*>dlm_tensor_ptr
+
+        # dl_tensor is identically filled for versioned and unversioned:
+        dl_tensor = &dlm_tensor_ver.dl_tensor
+        capsule_name = CAPSULE_NAME_VER
+
+        dlm_tensor_ver.manager_ctx = <void*>owner
+        dlm_tensor_ver.deleter = deleter_ver
+
+        # "Versioned" specific initialization:
+        dlm_tensor_ver.version.major = IMPL_VER_MAJOR
+        dlm_tensor_ver.version.minor = IMPL_VER_MINOR
+
+        # CuPy arrays are writeable but may be copied if copying to the CPU.
+        dlm_tensor_ver.flags = 0
+        if owner is not array:
+            dlm_tensor_ver.flags |= DLPACK_FLAG_BITMASK_IS_COPIED
+    else:
+        dlm_tensor = <DLManagedTensor*>dlm_tensor_ptr
+        # dl_tensor is identically filled for versioned and unversioned:
+        dl_tensor = &dlm_tensor.dl_tensor
+        capsule_name = CAPSULE_NAME
+
+        dlm_tensor.manager_ctx = <void*>owner
+        dlm_tensor.deleter = deleter
+
+    # Create capsule now. After the else, the capsule will clean up on error.
+    # (Note that it is good to handle expected BufferErrors early.)
+    try:
+        capsule = cpython.PyCapsule_New(
+            dlm_tensor_ptr, capsule_name, pycapsule_deleter)
+    except BaseException:
+        stdlib.free(dlm_tensor_ptr)
+        stdlib.free(shape_strides)
+        raise
+    else:
+        # Finalize dlm_tensor for `pycapsule_deleter`.  This else block must
+        # never fail/raise and initialize everything used by the deleter.
+        cpython.Py_INCREF(owner)
+        dl_tensor.shape = shape_strides
+
+    # And fill in all other fields (that are not part of the cleanup)
+    dl_tensor.ndim = ndim
+    dl_tensor.strides = shape_strides + ndim
+    dl_tensor.byte_offset = 0
+
+    dl_tensor.dtype.code = dtype_code
+    dl_tensor.dtype.lanes = <uint16_t>1
+    dl_tensor.dtype.bits = <uint8_t>(dtype_itemsize * 8)
+
+    dl_tensor.device = device
+
+    # Fill in the shape and strides information (depends on GPU vs CPU array).
+    # (assumes strides are a multiple of itemsize, that should be OK for cupy.)
+    if owner is array:
+        dl_tensor.data = <void *>array.data.ptr
```

Comment:
This is even more never-wrecking 😅
```suggestion
        dl_tensor.data = <void *><intptr_t>(array.data.ptr)
```
(Previously, the conversion is done through an intermediate assignment.)

Reply:
I think this is already `intptr_t` since it should stay in Cython.  But indeed no need to require the reader guess at that!